src/charm.py

```diff
@@ -851,6 +852,14 @@ def _handle_status_update_config(self, event: HookEvent) -> None:
             event.defer()
             return
 
+        if self.model.relations[KRATOS_INFO_RELATION_NAME] and self._public_url is None:
+            self.unit.status = BlockedStatus(
+                "Cannot send integration data without an external hostname. Please "
+                "provide an ingress relation or an external_url config."
+            )
+            event.defer()
+            return
```

Comment:
nit-pick: Maybe with this, you can undo the `raise_for_blocked` change in the tests.

```suggestion
        if self.model.relations[KRATOS_INFO_RELATION_NAME] and self.public_ingress.relation is None:
            self.unit.status = BlockedStatus(
                "Cannot send integration data without an external hostname. Please "
                "provide an ingress relation."
            )
            return
        elif self.model.relations[KRATOS_INFO_RELATION_NAME] and self._public_url is None:
            self.unit.status = WaitingStatus(
                "Waiting for ingress"
            )
            event.defer()
            return
```